GVRf/Framework/src/org/gearvrf/animation/keyframe/GVRNodeAnimationController.java

```diff
@@ -0,0 +1,86 @@
+package org.gearvrf.animation.keyframe;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import org.gearvrf.GVRSceneObject;
+import org.joml.Matrix4f;
+
+/**
+ * Controls node animation.
+ */
+public class GVRNodeAnimationController {
+    private static final String TAG = GVRNodeAnimationController.class.getSimpleName();
+
+    protected GVRSceneObject sceneRoot;
+    protected GVRKeyFrameAnimation animation;
+
+    protected class AnimationItem {
+        GVRSceneObject target;
+        int channelId;
+
+        AnimationItem(GVRSceneObject target, int channelId) {
+            this.target = target;
+            this.channelId = channelId;
+        }
+    }
+
+    protected List<AnimationItem> animatedNodes;
+
+    /**
+     * Constructs a list of animated {@link GVRSceneObject}.
+     *
+     * @param gvrContext The GVR context.
+     * @param sceneRoot The scene root.
+     * @param animation The animation object.
+     */
+    public GVRNodeAnimationController(GVRSceneObject sceneRoot, GVRKeyFrameAnimation animation) {
+        this.sceneRoot = sceneRoot;
+        this.animation = animation;
+
+        animatedNodes = new ArrayList<AnimationItem>();
+        if (animation != null) {
+            scanTree(sceneRoot);
+        }
+    }
+
+    /* Returns true if subtree contains renderables */
+    protected boolean scanTree(GVRSceneObject node) {
+        boolean containsRenderable = node.getRenderData() != null;
+
+        for (GVRSceneObject child : node.getChildren()) {
+            containsRenderable |= scanTree(child);
+        }
+
+        // Find channel Id
+        int channelId = animation.findChannel(node.getName());
+        if (channelId != -1 && containsRenderable) {
+            animatedNodes.add(new AnimationItem(node, channelId));
+        }
+
+        return containsRenderable;
+    }
+
+    /**
+     * Update node transforms at each animation step.
+     */
```

Comment:
It is possible to combine them but I chose to keep the skinning controller dedicated to only skinning. There are a couple of reasons:

1) The node animation code may be extended to support other types of animation: opacity / effects, etc. This can accidentally break the skinning controller or lower its performance

2) Both controllers have pruning to exclude nodes they don't control. So the total execution time is optimized, as they are one.

3) The animation transforms are in the animation class so there is no duplicate.